File: src/tasks/Statistics/Statistics.pl
#!/usr/bin/perl
#
# Basic statistical measurements
# includes:
#
#
# * min, max range
# * 1st, 2rd Quartile
# * median
# * Arithmetic mean
# * standard deviation
# * z-test implementation to find outliers
#

use warnings;
use strict;

my @unsorted = ();
open(F, "<", "data.txt") or die "Failed";
while(<F>) { chomp; push(@unsorted, int($_));}
close (F);
my $min;
my $max;
my $range;
my @array = sort {$a <=> $b} @unsorted;

# Arithmetic Mean  subroutine
sub my_mean{
	my $arraySize;
	my $sum = 0;
	for ( @array) {
	$sum += $_;
	}
	$arraySize = scalar(@array);
	$arraySize = $#array + 1;
	return my $mean = $sum / $arraySize;
}

# Min subroutine
sub my_min{
	for(@array){
		 $min = $_ if !$min || $_ < $min;
		}
	return $min;
}

# Max subroutine
sub my_max{
	for(@array){
		$max = $_ if !$max || $_ > $max;
}
return $max;
}

# Range subroutine
sub my_range{
	$range = my_max(@array) - my_min(@array);
	return $range;
}

# Median subroutine
sub my_median{
	my $med;
	if( (@array % 2) == 1 ) {
        $med = $array[((@array+1) / 2)-1];
    } else {
        $med = ($array[(@array / 2)-1] + $array[@array / 2]) / 2;
    }
    return $med;
}

# Standard Deviation subroutine
sub my_stdev{
	my $num;
	$num = scalar(@array);
	$num = $#array + 1;
	my $x;
	my $avg = my_mean(@array);
	my $sqsum = 0;
	foreach $x(@array){
	$sqsum += ($x - $avg)*($x - $avg) ;
	}
	my ($var)=$sqsum/$num;
	my ($stand_dev)=sqrt ($var);
	return $stand_dev;
}

# Z-test subroutine
sub my_ztest{
	my @ztest;
	my $avg = my_mean(@array);
	my $sd = my_stdev(@array);
	my $temp;
	foreach $_ (@array){
	$temp = abs( $_ - $avg)/ $sd;
	if ($temp > 3){
		push(@ztest, "$_");
	}
	}
	if (@ztest){
		return @ztest}
	else {
		return "No z values";
	}
	
}

# 1st quartile subroutine
sub my_firstquartile{
	my $med = my_median(@array);
	my @temp;
	foreach $_ (@array){
	if ($_ < $med){
		push(@temp, "$_");
	}
	}
	my $quart;
	my @vals = sort {$a <=> $b} @temp;
	if( (@vals % 2) == 1 ) {
        $quart = $vals[((@vals+1) / 2)-1];
    } else {
        $quart = ($vals[(@vals / 2)-1] + $vals[@vals / 2]) / 2;
    }
    return $quart;
}

# 3nd quartile subroutine
sub my_thirdquartile{
	my $med = my_median(@array);
	my @temp;
	foreach $_ (@array){
	if ($_ > $med){
		push(@temp, "$_");
	}
	}
	my $quart;
	my @vals = sort {$a <=> $b} @temp;
	if( (@vals % 2) == 1 ) {
        $quart = $vals[((@vals+1) / 2)-1];
    } else {
        $quart = ($vals[(@vals / 2)-1] + $vals[@vals / 2]) / 2;
    }
    return $quart;
}

print @array, "\n";

print "Average is : ",my_mean(@array), "\n";
print "Minimum value is:", my_min(@array),"\n";
print "Maximum value is:", my_max(@array),"\n";
print "Range between values is :", my_range(@array), "\n";
print "Median is: ", my_median(@array), "\n";
print "Standard deviation is: ", my_stdev(@array), "\n";
print "Z-test values are: ", my_ztest(@array), "\n";
print "1st quartile is : ", my_firstquartile(@array), "\n";
print "3nd quartile is : ", my_thirdquartile(@array), "\n";

File: src/tasks/Statistics/Statistics.js
<script>
	function loadFile(uri) {
		var r = new XMLHttpRequest();
		r.open('GET', uri, true);
		r.send(null);
		r.onreadystatechange = function() {
			if (r.readyState == 4) {
				processFile(r.responseText);
			}
		}
	}
	
	function processFile(fileContent) {
		var numbers = new Array();
		var lines = fileContent.split('\n');
		numbers = fileContent.split('\n');
		numbers.sort(function(a,b){return a-b});
		//alert(numbers); // sorted 
		for(var i=0; i < numbers.length; i++)
		{
			numbers[i] = parseInt(numbers[i]);
		}
		
		//Min
		var min = minimum(numbers);
		//Max
		var max = maximum(numbers);
		//Range
		var range = max - min;
		//Mean - Average
		var average = mymean(numbers);
		
		var quartile1 = quartiles(numbers, numbers.length, 1);
		var quartile2 = quartiles(numbers, numbers.length, 2);
		var quartile3 = quartiles(numbers, numbers.length, 3);
		
		var standardDev = stand_dev(numbers, numbers.length);
		var zTest = z_test(numbers, numbers.length);
		
		document.write('<p>Minimum value is : ' + min + '</p>' );
		document.write('<p>Maximum value is : ' + max + '</p>');
		document.write('<p>Range between values is : ' + range + '</p>');
		document.write('<p>Average value is : ' + average + '</p>');
		document.write('<p>First quartile is : ' + quartile1 + '</p>');
		document.write('<p>Median is : ' + quartile2 + '</p>');
		document.write('<p>Third quartile is : ' + quartile3 + '</p>');
		document.write('<p>The standard deviation is : ' + standardDev + '</p>');
		if (zTest.length > 0)
		{
			document.write('<p>Outliers are : ' + zTest + '</p>');
		}
		else
		{
			document.write('<p>No outliers'+ '</p>');
		}
		
		
	}
	
	function minimum(varArray)
	{
		var min =  Math.min.apply(Math,varArray);	
		return min;
	}
	
	function maximum(varArray)
	{
		var max =  Math.max.apply(Math,varArray);	
		return max;
	}
	
	function mymean(varArray)
	{
		var count_elements = varArray.length;
		var total_value = 0;
		for(var i=0; i < count_elements; i++)
		{
			total_value += varArray[i];
		}
		var average = parseFloat(total_value / count_elements);
		average = Number((average).toFixed(2));
		return average;
	}
	
	function quartiles(varArray, length, quartile){
		var median;
		var quartile_1;
		var quartile_3;
		var temp = length / 2;
		
		if(quartile == 1){
			if(length % 2 == 0){
				if(temp % 2 == 0){
					var numb = parseInt(temp / 2);
					var sum = varArray[numb - 1] + varArray[numb];
					quartile_1 = parseFloat(sum / 2);
					return quartile_1;
				}
				else{
					var numb = parseInt(temp / 2);
					quartile_1 = varArray[numb];
					return quartile_1;
				}
			}
			else if(length % 2 == 1){
				if(temp % 2 == 0){
					var numb = parseInt(temp / 2);
					var sum = varArray[numb - 1] + varArray[numb]
					quartile_1 = parseFloat(sum/2);
					return quartile_1;
				}
				else{
					var numb = parseInt(temp / 2);
					quartile_1 = varArray[numb];
					return quartile_1;
				}
			}
		}
		else if(quartile == 2){
			if(length % 2 == 1){
				var numb = parseInt(length / 2);
				median = varArray[numb];
				return median;
			}
			else{
				var numb = parseInt(length / 2);
				var element1 = varArray[numb -1 ];
				var element2 = varArray[numb];
				var sum = element1 + element2;
				median = parseFloat(sum/2);
				return median;
			}
		}
		else if(quartile == 3){
			if(length % 2 == 0){
				if(temp % 2 == 0){
					var numb = parseInt(temp / 2);
					var element1 = varArray[(3 * numb) - 1];
					var numb2 = parseInt((3 * temp) / 2);
					var element2 = varArray[numb2];
					var sum = element1 + element2;
					quartile_3 =  (sum/ 2);
					return quartile_3;
				}
				else{
					var numb = parseInt((3 * temp) / 2) ;
					quartile_3 = varArray[numb];
					return quartile_3;
				}
			}
			else if(length % 2 == 1){
				if(temp % 2 == 0){
					var numb = parseInt(temp / 2);
					var numb2 = parseInt((3 * temp) / 2);
					var numb3 = parseInt(((3 * temp) / 2) + 1);
					var sum = varArray[numb2] + varArray[numb3];
					quartile_3 = parseFloat(sum / 2);
					return quartile_3;
				}
				else{
					var numb = parseInt(temp / 2);
					var numb2 = parseInt(((3 * temp) / 2) + 1);
					quartile_3 = varArray[numb2];
					return quartile_3;
				}
			}
		}
	}
	
	function stand_dev(varArray, length){
		var sum = 0;
		var average = mymean(varArray, length);
		for(var i=0; i < length; i++)
		{
			var numb = parseFloat(varArray[i]- average);
			//numb = Math.abs(numb);
			var test = Math.pow(numb,2);
			sum +=  Math.pow(numb,2);
		}
		var squareRoot = parseFloat(sum / length);
		stdev = Math.sqrt(squareRoot);
		var result = Number((stdev).toFixed(3));
		return result;
	}

	function z_test(varArray, length){
	
		var temp = 0;
		var z_array = new Array();
		var average = mymean(varArray, length);
		var stdev = stand_dev(varArray, length);
		//calculate numbers out of boundaries
		for(var i=0; i < length; i++)
		{
			var numb = Math.abs(varArray[i] - average);
			var gin = parseFloat(numb / stdev );
			if( gin > 3)
			{
				temp += 1;
			}
		}
		
		if( temp == 0 ){
			return 0;
		}
		else{
			// fill array with numbers out of boundaries
			for(var j=0; j < length; j++){
				var numb2 = Math.abs(varArray[j] - average);
				var gin2 = parseFloat(numb2 / stdev );
				
				if(gin2 > 3){
					for( var k = 0;  k < temp; k++){
					//alert(varArray[k]);
					z_array.push(varArray[j]);
					}
				}
			}	
			return z_array;
		}
	}

	loadFile('http://localhost/Statistics/data.txt');
	</script>